Priority Queue/SimplePriorityQueue.cs

```diff
@@ -355,13 +409,31 @@ public bool TryRemove(TItem item)
         {
             lock(_queue)
             {
-                SimpleNode removeMe = GetExistingNode(item);
-                if(removeMe == null)
+                SimpleNode removeMe;
+                IList<SimpleNode> nodes;
+                if (item == null)
                 {
-                    return false;
+                    if (_nullNodesCache.Count == 0)
+                    {
+                        return false;
+                    }
+                    removeMe = _nullNodesCache[0];
+                    nodes = _nullNodesCache;
+                }
+                else
+                {
+                    if (!_itemToNodesCache.TryGetValue(item, out nodes))
+                    {
+                        return false;
+                    }
+                    removeMe = nodes[0];
+                    if (nodes.Count == 1)
+                    {
+                        _itemToNodesCache.Remove(item);
+                    }
                 }
                 _queue.Remove(removeMe);
-                RemoveFromNodeCache(removeMe);
+                nodes.Remove(removeMe);
                 return true;
             }
         }
```

Comment:
I'm not sure how I feel about a lot of the changes to `SimplePriorityQueue`.  It's meant for cases where performance is not a concern - if they need the best performance possible, they should use `FastPriorityQueue`.  I'd prefer to avoid optimizations that increase the code complexity/size if at all possible.